### ISSUE TITLE
SMTPClient.cpp - Incomplete error logging

### ISSUE THREAD

**Issue by TimelessNL** [domoticz/domoticz#1077]
Hi,

**Small introduction:**
I was working on a OpenWRT/LEDE (Project) Domoticz build which was working fine for now. I accidentally left out the libary extension SMTP/SMTPS for libcurl. From where ofcourse sending mails failed. But the errors I received in the logging were incomplete and missed the point.

**Example/Problem:**
```
2016-12-27 21:39:49.321 Error: SMTP Mailer: Error sending Email to: !
2016-12-27 21:39:49.322 Error: Failed to send email notification :
2016-12-27 21:39:49.322 Error: Notification sent (email) => Failed
```

It seems that the "m_Recipients" array is empty, at least at [0]. But for some reason `if (m_Recipients.size()==0)` is not matched but still `m_Recipients[0].c_str().` gives nothing between `Email to: !`. It could be that it was printing the `_log.Log(LOG_ERROR,"SMTP Mailer: Error sending Email to: %s !",m_Recipients[0].c_str());` in the catch block.

I first taught that this had to do with me missing the SMTP/SMTPS extension. But after enabling the extension, it still showed these messages when I for example entered a wrong password. So there seems to be something wrong with the printout. (Sending mails worked perfectly with correct username/password)

_Also as a side note, I think we should lose the "exclamation mark" after the recipients address. Because now it seems that it tries to send mails to a "exclamation mark". The "exclamation mark" does not add anything to the context and is only confusing._

A quick search on the forums reveiled that I was not the only one seeing these kind of log errors:
[Sending EMail fails](https://www.domoticz.com/forum/viewtopic.php?t=12315#p88352)
[smtp error](https://www.domoticz.com/forum/viewtopic.php?t=14239#p103986)
So I don't think that is has something to do with my OpenWRT/LEDE builds. I do not have a fix for this unfortunately. I only noticed this kind of behavior. So lets discuss if it is worth fixing.

----------------------------------

**Nice to have:**
Since Domoticz compiled fine while missing the libcurl SMTP/SMTPS extension. Is there a way to detect if the extension is missing while sending mails (and log if fails to detect?) or maybe even during compile time?

And maybe differentiate the Catch Block(unplanned) logging and the normal error(planned) logging.

**Comment by gizmocuz** [comment]
the "exclamation mark" is used in all error logs, like ERROR... sure there needs to be something before this, like an email address
It looks your recipient address is empty, check your script/blockly
As for the second log line, this was defined wrong by a developer, in the next update it will just read
'Failed to send Email notification!'

As for the nice to have, i could not find it, i have seen a function for PHP (get loaded extensions), but not for c++

**Comment by TimelessNL** [comment]
Then maybe print "empty" or "none" where the mail address should be, instead of a empty string? This makes it more clear. Because sure there should be a mail address, but something caused it to not be there.

Let me explain my configuration a little bit better: I'm using the "test" button in the settings->email settion after configuring (and saved) all the required mail settings. But when either I enter  a wrong password I see this error message without a email address in the logs after I pressed the "test" button. So I'm not using any script/blocky stuff. But if my recipient address was empty, this would have matched the if satement I showed earlier. But it doesn't, which seems to indicate that the size is not 0.

And for the 2nd log line. Would it be possible to print the exception and/or exception code? This makes debugging and solving issues on the forum easier.

And for the nice to have. Maybe we could detect the exception, and catch specific exceptions that have something to do with loading the library?

[Send from my mobile phone]
